src/vehicle/vehicle.go

```diff
@@ -1,126 +1,172 @@
+	pipe, err := vehicle.getPipe(order.ID)
+
+	if err != nil {
+		vehicle.trace.Error().Err(err).Msg("error getting pipe")
 		return err
 	}
 
-	valuesMap := fieldsValuesToMap(order.Fields)
-	valuesMap = vehicle.displayConverter.Revert(valuesMap)
-	valuesMap = vehicle.podConverter.Convert(valuesMap)
-
-	valuesRaw := vehicle.parser.Encode(order.ID, valuesMap)
-	bitArray := vehicle.parser.CreateBitArray(order.ID, fieldsEnableToMap(order.Fields))
-	fullRaw := append(valuesRaw, bitArray...)
-	log.Println("fullRaw", fullRaw)
+	values := getOrderValues(order, vehicle.trace)
+	convertedValues := vehicle.applyUnitConversion(values)
 
-	_, err := common.WriteAll(pipe, fullRaw)
+	buf := new(bytes.Buffer)
 
-	vehicle.statsMx.Lock()
-	defer vehicle.statsMx.Unlock()
+	idBuf := make([]byte, 2)
+	binary.LittleEndian.PutUint16(idBuf, order.ID)
 
-	if err == nil {
-		vehicle.stats.sent++
-	} else {
-		vehicle.trace.Error().Stack().Err(err).Msg("")
-		vehicle.stats.sentFail++
+	err = vehicle.packetParser.Encode(order.ID, convertedValues, buf)
+	if err != nil {
+		vehicle.trace.Error().Err(err).Msg("error encoding order")
+		return err
 	}
 
+	fullBuf := append(idBuf, buf.Bytes()...)
+
+	_, err = common.WriteAll(pipe, fullBuf)
+
 	return err
 }
 
-func fieldsValuesToMap(fields map[string]models.Field) map[string]any {
-	valuesMap := make(map[string]any)
+func (vehicle *Vehicle) applyUnitConversion(values map[string]packet.Value) map[string]packet.Value {
+	newValues := make(map[string]packet.Value)
+
+	for name, value := range values {
+		switch typedValue := value.(type) {
+		case packet.Numeric:
+			valueInSIUnits, podErr := vehicle.podConverter.Revert(name, float64(typedValue))
+
+			if podErr != nil {
+				vehicle.trace.Error().Err(podErr).Msg("error reverting podUnits")
+			}
+
+			valueInDisplayUnits, displayErr := vehicle.displayConverter.Convert(name, valueInSIUnits)
 
-	for name, field := range fields {
-		valuesMap[name] = field.Value
+			if displayErr != nil {
+				vehicle.trace.Error().Err(displayErr).Msg("error converting to displayUnits")
+
+			}
+
+			newValues[name] = packet.Numeric(valueInDisplayUnits)
+		default:
+			newValues[name] = typedValue
+		}
+	}
+
+	return newValues
+}
+
+func (vehicle *Vehicle) getPipe(id uint16) (*pipe.Pipe, error) {
+	board, ok := vehicle.idToBoard[id]
+	if !ok {
+		return nil, fmt.Errorf("board for id %d not found", id)
+	}
+
+	pipe, ok := vehicle.pipes[board]
+	if !ok {
+		return nil, fmt.Errorf("pipe for board %s not found", board)
+	}
+
+	return pipe, nil
+}
+
+func getOrderValues(order models.Order, trace zerolog.Logger) map[string]packet.Value {
+	values := make(map[string]packet.Value)
+
+	for name, field := range order.Fields {
+		switch value := field.Value.(type) {
+		case float64:
+			values[name] = packet.Numeric(value)
+		case bool:
+			values[name] = packet.Boolean(value)
+		case string:
+			values[name] = packet.Enum(value)
+		default:
+			trace.Error().Msg("order field value not recognized")
```

Comment:
store the type (with `.Str("type", fmt.Sprintf("%T", field.Value))`) and the name (with `.Str("name", name)`)